### PR TITLE
task: various small improvements to `LocalKey`

### PR DESCRIPTION
**Pull request by Darksonn**
Fixes the following small issues in `LocalKey`:

 * Describe all situations in which the methods can panic.
 * Also set the task-local when running the destructor of the future.
 * Don't panic in `try_with` if the thread-local holding the task-local is destroyed. (as part of shutting down the thread)
 * Improve panic error messages.
 * Implement `Debug` for `TaskLocalFuture`.

### CODE REVIEW

**Review by LucioFranco — APPROVED** [review]
LGTM, I didn't review the drop part super hard since I don't really remember the pin and drop interactions anymore but I assume its fine, the 'static bound should make it all legal anyways iirc from when we designed this.

File: tokio/src/task/task_local.rs [hunk 1]
```diff
@@ -161,34 +175,85 @@ impl<T: 'static> LocalKey<T> {
...
-        crate::pin!(scope);
-        scope.with_task(|_| f())
+        let mut value = Some(value);
+        match self.scope_inner(&mut value, f) {
+            Ok(res) => res,
+            Err(ScopeInnerErr::BorrowError) => {
+                panic!("sync_scope called while Task Local Storage is borrowed")
+            }
+            Err(ScopeInnerErr::AccessError) => {
+                panic!("cannot access a Task Local Storage value during or after destruction")
```
Comment (hawkw) [inline tokio/src/task/task_local.rs:193]: very tiny nit: in the docs, we write "task-local storage", but in the panics, we write "Task Local Storage". i have a slight preference for keeping this consistent:
```suggestion
                panic!("sync_scope called while task-local storage is borrowed")
            }
            Err(ScopeInnerErr::AccessError) => {
                panic!("cannot access a task-local storage value during or after destruction")
```

File: tokio/src/task/task_local.rs [hunk 2]
```diff
@@ -161,34 +175,85 @@ impl<T: 'static> LocalKey<T> {
...
-            future: (),
-            _pinned: PhantomPinned,
-        };
-        crate::pin!(scope);
-        scope.with_task(|_| f())
+        let mut value = Some(value);
+        match self.scope_inner(&mut value, f) {
+            Ok(res) => res,
+            Err(ScopeInnerErr::BorrowError) => {
+                panic!("sync_scope called while Task Local Storage is borrowed")
```
Comment (hawkw) [inline tokio/src/task/task_local.rs:190]: maybe:
```suggestion
                panic!("`LocalKey::sync_scope` called while Task Local Storage is borrowed")
```

File: tokio/src/task/task_local.rs [hunk 3]
```diff
@@ -224,76 +301,109 @@ impl<T: 'static> fmt::Debug for LocalKey<T> {
...
-        };
+        match res {
+            Ok(Some(res)) => res,
+            Ok(None) => panic!("TaskLocalFuture polled after completion"),
+            Err(ScopeInnerErr::BorrowError) => {
+                panic!("TaskLocalFuture::poll called while task local is borrowed")
+            }
+            Err(ScopeInnerErr::AccessError) => {
+                panic!("cannot access a Task Local Storage value during or after destruction")
+            }
```
Comment (hawkw) [inline tokio/src/task/task_local.rs:367]: nits:
```suggestion
            Ok(None) => panic!("`TaskLocalFuture` polled after completion"),
            Err(ScopeInnerErr::BorrowError) => {
                panic!("`TaskLocalFuture::poll` called while task-local storage is borrowed")
            }
            Err(ScopeInnerErr::AccessError) => {
                panic!("cannot access a task-local storage value during or after destruction")
            }
```

File: tokio/src/task/task_local.rs [hunk 4]
```diff
@@ -161,34 +175,85 @@ impl<T: 'static> LocalKey<T> {
...
     where
         F: FnOnce(&T) -> R,
     {
-        self.try_with(f).expect(
-            "cannot access a Task Local Storage value \
-             without setting it via `LocalKey::set`",
-        )
+        match self.try_with(f) {
+            Ok(res) => res,
+            Err(_) => panic!("cannot access a Task Local Storage value without setting it first"),
```
Comment (hawkw) [inline tokio/src/task/task_local.rs:255]: nit: maybe
```suggestion
            Err(_) => panic!("cannot access a task-local storage value without setting it first"),
```

File: tokio/src/task/task_local.rs [hunk 5]
```diff
@@ -161,34 +175,85 @@ impl<T: 'static> LocalKey<T> {
...
+                // However, we never give user-code access to the guards, so
+                // there's no way for user-code to forget to destroy a guard.
+                //
+                // The call to `with` also should not panic, since the
+                // thread-local wasn't destroyed when we first called
+                // `scope_inner`, and it shouldn't have gotten destroyed since
+                // then.
+                self.local.inner.with(|inner| {
+                    let mut ref_mut = inner.borrow_mut();
+                    std::mem::swap(self.slot, &mut *ref_mut);
```
Comment (hawkw) [inline tokio/src/task/task_local.rs:223]: tiny style nit: we use multiple `std::mem` functions in this file (`swap` and `needs_drop`), i might consider a `use std::mem;`

**Review by hawkw — APPROVED** [review]
a handful of nits, but overall this lgtm!

File: tokio/src/task/task_local.rs [hunk 6]
```diff
@@ -224,76 +301,109 @@ impl<T: 'static> fmt::Debug for LocalKey<T> {
...
-            local: *project.local,
-        };
+        match res {
+            Ok(Some(res)) => res,
+            Ok(None) => panic!("TaskLocalFuture polled after completion"),
+            Err(ScopeInnerErr::BorrowError) => {
+                panic!("TaskLocalFuture::poll called while task local is borrowed")
+            }
+            Err(ScopeInnerErr::AccessError) => {
+                panic!("cannot access a Task Local Storage value during or after destruction")
```
Comment (hawkw) [inline tokio/src/task/task_local.rs:366]: tiny nit: this panic message shows up in a couple places, maybe it should be a const?

File: tokio/src/task/task_local.rs [hunk 7]
```diff
@@ -161,34 +175,85 @@ impl<T: 'static> LocalKey<T> {
...
+            slot: &'a mut Option<T>,
+        }
+
+        impl<'a, T: 'static> Drop for Guard<'a, T> {
+            fn drop(&mut self) {
+                // This should not panic.
+                //
+                // We know that the RefCell was not borrowed before the call to
+                // `scope_inner`, so the only way for this to panic is if the
+                // closure has created but not destroyed a RefCell guard.
```
[the remark this answers is not in the document]
Comment (Darksonn) [reply tokio/src/task/task_local.rs:213]: Actually, this should be "has". It's an action that the closure may or may not have taken.

**Pull request merged**